rosbridge_library/src/rosbridge_library/internal/ros_loader.py

```diff
@@ -210,6 +223,9 @@ def _splittype(typestring):
         return (splits[0], splits[2])
     if len(splits) == 2:
         return (splits[0], splits[1])
+    if len(splits) == 4 and splits[1] == "action":
+        # Special case for hidden services that are autogenerated for actions
+        return (splits[0], splits[3])
```

Comment:
I think this is the only required change to allow loading hidden service definitions. If I do that, I can successfully load a action service from the client.

```
echo '{ "op": "advertise_service", "type": "test_msgs/action/_nested_message/NestedMessage_SendGoal", "service": "foo" }' | websocat ws://localhost:9090 --no-close
```

Note however, that the client needs to specify the explicit path of the service (`test_msgs/action/NestedMessage_SendGoal` would not work), but I think this is acceptable for now?

---

We could probably combine the logic and write it simpler like so:
```
    if len(splits) >= 3:
        return (splits[0], splits[-1])
```

Reply:
I think for advertise service this is easier since the client (roslibjs in our case) needs to specify a type and the AdvertiseService Capability takes the type into consideration.

---

For calling a service however the type is automatically determined by the rclpy implementation and a given type argument would get ignored (https://github.com/RobotWebTools/rosbridge_suite/blob/ros2/rosbridge_library/src/rosbridge_library/internal/services.py#L109) . So it should either be optional to specify a type for service calls or a way to work around the rcl issue. (Best case would be that it gets fixed in rclpy or upstream, I couldn't pinpoint the issue yet)

Pros for making it optional for the call_service operation as well:
- the implementation won't implicitly fix up/modify the type after determination
- decreased complexity in guarding this one special case
- decreased complexity to automatically determine the "_snake_cased" service type

One big con: 
- the type argument is not part of the Protocol for call_service: https://github.com/RobotWebTools/rosbridge_suite/blob/ros2/ROSBRIDGE_PROTOCOL.md#336-call-service


Also (but this shouldn't affect any implementation) a lot of the current test cases cover things like "////invalid_name" where with 
```
if len(splits >=2):  # taking your suggestion one step further
    return (splits[0], splits[-1])
```

all of those won't raise InvalidTypeStringException anymore and another way to determine an invalid service name is needed

Reply:
Good point, make sense.

Best would be indeed if we could fix the underlying `rclpy/rcl` issue. 
I'm not very familiar with ROS2 actions, but I quickly tried running the [example action server](https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Actions/Understanding-ROS2-Actions.html#setup) but `ros2 service --include-hidden-services list -t` does not list any action related services :thinking:

Reply:
It doesn't? Thats strange. For me running `ros2 run action_tutorials_cpp fibonacci_action_server` in one terminal and in another one `ros2 service --include-hidden-services list -t` yields 
```
/fibonacci/_action/cancel_goal [action_msgs/srv/CancelGoal]
/fibonacci/_action/get_result [action_tutorials_interfaces/action/Fibonacci_GetResult]
/fibonacci/_action/send_goal [action_tutorials_interfaces/action/Fibonacci_SendGoal]
/fibonacci_action_server/describe_parameters [rcl_interfaces/srv/DescribeParameters]
/fibonacci_action_server/get_parameter_types [rcl_interfaces/srv/GetParameterTypes]
/fibonacci_action_server/get_parameters [rcl_interfaces/srv/GetParameters]
/fibonacci_action_server/list_parameters [rcl_interfaces/srv/ListParameters]
/fibonacci_action_server/set_parameters [rcl_interfaces/srv/SetParameters]
/fibonacci_action_server/set_parameters_atomically [rcl_interfaces/srv/SetParametersAtomically]
```


(excerpt from `ros2 doctor --report`:
```
   PLATFORM INFORMATION
system           : Linux
platform info    : Linux-5.4.0-126-generic-x86_64-with-glibc2.29
release          : 5.4.0-126-generic
processor        : x86_64

   RMW MIDDLEWARE
middleware name    : rmw_fastrtps_cpp

   ROS 2 INFORMATION
distribution name      : foxy
distribution type      : ros2
distribution status    : active
release platforms      : {'ubuntu': ['focal']}

```

Reply:
It might be an issue/design inconsistency/faulty expectation with how rosidl_python creates the python implementations for the action types. I will check with rclcpp and link the resulting issue in this PR as well. Especially if the workaround is necessary for humble as well (might just another reason to upgrade)